### PR TITLE
Bodrov Daniil. Lab 2. Var 1. Instrument Function Pass

### PR DESCRIPTION
**Pull request by overinvest**
This pull request implements a tool that instruments all functions in a module by adding `instrument_start()` calls at the beginning of each function and `instrument_end()` calls at the end of each function.

The program works by traversing all functions in the module and inserting instrumentation calls at the necessary locations. Important lines of code include the `run()` function, which performs the instrumentation of functions, and the `isInstrumentedCall()` and `insertInstrumentCall()` functions, which check for and insert instrumentation calls, respectively.

Additionally, the pull request adds information about the plugin and its version to easily embed the instrument into an existing process using the LLVM plugin linker.

Overall, the goal of this code is to instrument all functions in a module by automatically adding instrumentation calls at the start and end of each function without manual intervention.

`test.cpp`
```
int calculate_x(int num) {

    num += num + 1;
    num *= num;

    return num;
}

void func() {
    return;
}

double clamp(double v, double lo, double hi) {
  if (v > hi) {
    return hi;
  } else if (v < lo) {
    return lo;
  }
  return v;
}
```

`test.ll`

```
define dso_local noundef i32 @_Z11calculate_xi(i32 noundef %num) #0 {
entry:
  call void @instrument_start()
  %num.addr = alloca i32, align 4
  store i32 %num, ptr %num.addr, align 4
  %0 = load i32, ptr %num.addr, align 4
  %add = add nsw i32 %0, 1
  %1 = load i32, ptr %num.addr, align 4
  %add1 = add nsw i32 %1, %add
  store i32 %add1, ptr %num.addr, align 4
  %2 = load i32, ptr %num.addr, align 4
  %3 = load i32, ptr %num.addr, align 4
  %mul = mul nsw i32 %3, %2
  store i32 %mul, ptr %num.addr, align 4
  %4 = load i32, ptr %num.addr, align 4
  call void @instrument_end()
  ret i32 %4
}

define dso_local void @_Z4funcv() #0 {
entry:
  call void @instrument_start()
  call void @instrument_end()
  ret void
}

define dso_local noundef double @_Z5clampddd(double noundef %v, double noundef %lo, double noundef %hi) #0 {
entry:
  call void @instrument_start()
  %retval = alloca double, align 8
  %v.addr = alloca double, align 8
  %lo.addr = alloca double, align 8
  %hi.addr = alloca double, align 8
  store double %v, ptr %v.addr, align 8
  store double %lo, ptr %lo.addr, align 8
  store double %hi, ptr %hi.addr, align 8
  %0 = load double, ptr %v.addr, align 8
  %1 = load double, ptr %hi.addr, align 8
  %cmp = fcmp ogt double %0, %1
  br i1 %cmp, label %if.then, label %if.else

if.then:                                          ; preds = %entry
  %2 = load double, ptr %hi.addr, align 8
  store double %2, ptr %retval, align 8
  br label %return

if.else:                                          ; preds = %entry
  %3 = load double, ptr %v.addr, align 8
  %4 = load double, ptr %lo.addr, align 8
  %cmp1 = fcmp olt double %3, %4
  br i1 %cmp1, label %if.then2, label %if.end

if.then2:                                         ; preds = %if.else
  %5 = load double, ptr %lo.addr, align 8
  store double %5, ptr %retval, align 8
  br label %return

if.end:                                           ; preds = %if.else
  br label %if.end3

if.end3:                                          ; preds = %if.end
  %6 = load double, ptr %v.addr, align 8
  store double %6, ptr %retval, align 8
  br label %return

return:                                           ; preds = %if.end3, %if.then2, %if.then
  %7 = load double, ptr %retval, align 8
  call void @instrument_end()
  ret double %7
}

declare void @instrument_start()
declare void @instrument_end()
```

### CODE REVIEW

File: llvm/labs/lab2/bodrov_daniil/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+set(Group "FIIT1")
+set(PluginName "${Title}_${Student}_${Group}")
+
+# Result plugin name: Instrument_Functions_Pass_Bodrov_Daniil_FIIT1
+
+if (NOT WIN32 AND NOT CYGWIN)
+    message("-- ${PluginName} -- BUILD")
+
+    file(GLOB_RECURSE ALL_SOURCE_FILES *.cpp *.h *.hpp)
+    add_llvm_pass_plugin(${PluginName} ${ALL_SOURCE_FILES})
```
Comment (m-ly4) [inline llvm/labs/lab2/bodrov_daniil/CMakeLists.txt]:
```suggestion
    add_llvm_pass_plugin(${PluginName} ${ALL_SOURCE_FILES}
    DEPENDS intrinsics_gen BUILDTREE_ONLY)
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/bodrov_daniil/instrument_functions_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        module->getOrInsertFunction("instrument_end", funcType);
+
+    // Check if instrument_start() or instrument_end() has been already inserted
+    bool startInserted = false;
+    bool endInserted = false;
+
+    // Check for existing instrument calls in the function
+    for (auto &block : F) {
+      for (auto &instruction : block) {
+        if (isInstrumentedCall(instruction, startFunc)) {
```
Comment (m-ly4) [inline llvm/labs/lab2/bodrov_daniil/instrument_functions_pass.cpp]: Please write the check via searching over users of instrument_start/instrument_end functions.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/bodrov_daniil/instrument_functions_pass.cpp (hunk 2 above)
Comment (overinvest) [reply llvm/labs/lab2/bodrov_daniil/instrument_functions_pass.cpp]: Like that?

```
for (auto *user : startFunc.getCallee()->users()) {
      if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(user)) {
        if (callInst->getParent()->getParent() == &F) {
          startInserted = true;
          break;
        }
      }
    }

    for (auto *user : endFunc.getCallee()->users()) {
      if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(user)) {
        if (callInst->getParent()->getParent() == &F) {
          endInserted = true;
          break;
        }
      }
    }
```

Comment (m-ly4) [reply llvm/labs/lab2/bodrov_daniil/instrument_functions_pass.cpp]: Yes, like that. You also can put common code of this check into a separate function to reduce duplication.

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**